lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -13,548 +13,1055 @@
 
-/// Convert an Annotation into an optional AugmentedBundleType.  Returns None if
-/// the annotation is not an AugmentedBundleType.
-static Optional<AugmentedBundleType> decodeBundleType(Annotation anno) {
-  auto defName = anno.getMember<StringAttr>("defName");
-  auto elements = anno.getMember<ArrayAttr>("elements");
-  if (!defName || !elements)
-    return {};
-  AugmentedBundleType bundle(
-      {defName.getValue(), SmallVector<AugmentedField>()});
-  for (auto element : elements) {
-    auto field = decodeField(element);
-    if (!field)
-      return {};
-    bundle.elements.push_back(field.getValue());
+  Ground(DictionaryAttr dict) : AugmentedType(Kind::Ground, dict) {}
+
+  static bool classof(const AugmentedType *at) {
+    return at->getKind() == Kind::Ground;
   }
-  return Optional<AugmentedBundleType>(bundle);
-}
+};
 
-/// Remove Grand Central Annotations associated with SystemVerilog interfaces
-/// that should emitted.  This pass works in three major phases:
-///
-/// 1. The circuit's annotations are examnined to figure out _what_ interfaces
-///    there are.  This includes information about the name of the interface
-///    ("defName") and each of the elements (sv::InterfaceSignalOp) that make up
-///    the interface.  However, no information about the _type_ of the elements
-///    is known.
+/// Template used for representing any "unsupported" AugmentedType
+template <AugmentedType::Kind k>
+struct Unsupported : public AugmentedType {
+  StringAttr getClass() { return storage.getAs<StringAttr>("class"); }
+  DictionaryAttr getUnderlying() { return storage; }
+
+  Unsupported(DictionaryAttr dict) : AugmentedType(k, dict) {}
+
+  static bool classof(const AugmentedType *at) { return at->getKind() == k; }
+};
+
+/// A wrapper around a string that is used to encode a type which cannot be
+/// represented by an mlir::Type for some reason.  This is currently used to
+/// represent either an interface, a n-dimensional vector of interfaces, or a
+/// tombstone for an actually unsupported type (e.g., an AugmentedBooleanType).
+struct VerbatimType {
+  /// The textual representation of the type.
+  SmallString<0> str;
+
+  /// True if this is a type which must be "instatiated" and requires a trailing
+  /// "()".
+  bool instantiation;
+
+  /// Serialize this type to a string.
+  std::string toStr() {
+    if (instantiation)
+      return (str + "();").str();
+    return (str + ";").str();
+  }
+};
+
+/// A sum type representing either a type encoded as a string (VerbatimType)
+/// or an actual mlir::Type.
+typedef std::variant<VerbatimType, Type> TypeSum;
+
+/// A namespace that is used to store existing names and generate names.  This
+/// exists to work around limitations of SymbolTables.
+struct CircuitNamespace {
+
+private:
+  StringSet<> internal;
+  CircuitNamespace(StringSet<> internal) : internal(internal) {}
+
+public:
+  CircuitNamespace(CircuitOp op) {
+    op.walk([&](Operation *op) {
```

Comment:
This seems exactly what SymbolTable does.

Reply:
I didn't look, but I expect you're correct. This whole thing really just exists so you can insert strings into the namespace and avoid the checking that `SymbolTable` does about ops in the symbol table having no intermediary parent ops. 🤷‍♀️